keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,110 @@
+package io.improbable.keanu.algorithms.sampling;
+
+import com.google.common.collect.Sets;
+import io.improbable.keanu.KeanuRandom;
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.algorithms.PosteriorSamplingAlgorithm;
+import io.improbable.keanu.algorithms.ProbabilisticModel;
+import io.improbable.keanu.algorithms.Variable;
+import io.improbable.keanu.algorithms.graphtraversal.TopologicalSort;
+import io.improbable.keanu.algorithms.mcmc.NetworkSamplesGenerator;
+import io.improbable.keanu.algorithms.mcmc.SamplingAlgorithm;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.util.status.StatusBar;
+import io.improbable.keanu.vertices.Probabilistic;
+import io.improbable.keanu.vertices.Vertex;
+import org.nd4j.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class Forward implements PosteriorSamplingAlgorithm {
+
+    private final KeanuRandom random;
+
+    public Forward() {
+        this(KeanuRandom.getDefaultRandom());
+    }
+
+    public Forward(KeanuRandom random) {
+        this.random = random;
+    }
+
+    /**
+     * Samples from the prior of a Probabilistic Model.
+     * Bayesian Network is the only supported implementation.
+     *
+     * Samples are taken by sampling from the prior of the desired variables in topological order.
+     *
+     * @param model the model to sample from
+     * @param variablesToSampleFrom the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        Variable latent = latentVariables.get(0);
+        Preconditions.checkArgument(latent instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex.");
```

Comment:
Is it sufficient to check only the first variable? (ie if the first is a vertex are we guaranteed that all will be, and that this will always be true?).

Reply:
Fair point - will check them all